utils/src/main/java/software/amazon/awssdk/utils/Either.java

```diff
@@ -103,5 +103,49 @@ public void apply(Consumer<? super L> lFunc, Consumer<? super R> rFunc) {
     public static <L, R> Either<L, R> right(R value) {
         return new Either<>(Optional.empty(), Optional.of(value));
     }
+
+    /**
+     * Create a new Either from two possibly null values.
+     *
+     * If both values are null, null is returned. Only one of the left or right values
+     * must be non-null, otherwise an {@link IllegalArgumentException} is thrown.
+     * @param left The left value (possibly null)
+     * @param right The right value (possibly null)
+     * @param <L> Left type
+     * @param <R> Right type
+     * @return an Either representing one of the two values.
+     */
+    public static <L, R> Either<L, R> fromNullable(L left, R right) {
```

Comment:
Should this just return the Optional?

Reply:
I considered that, but it felt like a bit of a smell - it didn't feel like it was this things responsibility to determine how to handle `null`, `null` - I didn't want to wrap it in `Optional` if the caller was just going to throw an exception in the case of `null` it seemed like an unnecessary object.

Now both our use-cases of this class *do* immediately wrap in `Optional` so there's definitely a case to be made - I just felt that this was more in line with SRP.

Reply:
But isn't Optional the new type safe null :)

Reply:
*sigh* i suppose it is. It's another object tho. I don't really care enough to fight about it though - I'd rather get the PR though and remove some duplication - so I'll update it.

Reply:
I agree with Andrew, IMHO, returning null seems error-prune to me and it expects the caller to check before use. If the caller misses the null check, there's gonna be a runtime NPE. Side note, returning null is also not recommended in *Clean Code* :)